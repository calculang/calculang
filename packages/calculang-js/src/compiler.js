/*
Copyright (C) 2021  DCN Consulting Ltd (incorporated in Ireland)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License version 3,
as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

// compiler is implemented as a call to webpack v4 (atm) api with a loader setup to transform source code (via babel with a custom loader: calculang-transform-loader).
// The output is a Javascript bundle and sourcemap created by webpack, according to configuration below.
// pre-processed calculang information is needed to complete the compile, this comes from the introspection api call below, where results are passed to calculang-transform-loader via options.

import path from 'path';
import webpack from 'webpack';
import { createFsFromVolume, Volume } from 'memfs';

import introspection from '@calculang/introspection-api';

export default async (entrypoint, options = {}) => {
  const introspection_info = await introspection(entrypoint, options);
  // now pass this information to loader via options in loader/use config:

  var use_config = [
    {
      loader: path.resolve(__dirname, './calculang-transform-loader/index.js'),
      options: { ...options, outputLocation: '', ...introspection_info },
    },
  ];
  if (options.memo)
    // this is opt-in atm
    use_config.push({
      loader: path.resolve(
        __dirname,
        '../../introspection-api/dist/memoloader.js' // TODO better way to reference this?
      ),
      options,
    });

  // run webpack code
  const compiler = webpack({
    context: __dirname,
    entry: `${path.resolve(entrypoint)}`, // any way to get this running relative?
    devtool: 'source-map',
    output: {
      //path: path.resolve(__dirname),
      filename: path.basename(entrypoint, '.cul.js') + '.js', // gets rid of .cul
      libraryTarget: 'umd', // for most of dev was commonjs2, but didn't work well for browser? And I don't want sep. browser/nodejs bundles. Is ESM an eventual solution here?
      globalObject: 'this', // for UMD in browser,node

      // https://github.com/webpack/webpack/issues/3603#issuecomment-357664819 useful discussion
      devtoolModuleFilenameTemplate(info) {
        return `webpack:///${path
          .relative(__dirname, info.absoluteResourcePath)
          .replace(/\\/g, '/')}`; // do I need to add info.loaders?
      },
    },
    optimization: {
      minimize: false,
      concatenateModules: false, // Do this temporarily because concatenated modules leaks absolute paths - TODO fix!
      //runtimeChunk: true, results in a 0.bundle.js without bootstrap, just model code bundled as jsonP
    },
    module: {
      rules: [
        {
          test: /\.cul/,
          use: use_config,
        },
        {
          test: /\.sl/, // this was an earlier development extension (seed-lang!), keeping configured, because I still want to run my .sl files for a little while
          use: use_config,
        },
      ],
    },
  });

  compiler.outputFileSystem = createFsFromVolume(new Volume()); // will it be better to use disk directly?
  compiler.outputFileSystem.join = path.join.bind(path);

  return new Promise((resolve, reject) => {
    compiler.run((err, stats) => {
      if (err) reject(err);
      if (stats.hasErrors()) reject(stats.toJson().errors);

      //console.log(Object.keys(stats.compilation.assets));

      resolve({
        bundle:
          stats.compilation.assets[
            path.basename(entrypoint, '.cul.js') + '.js'
          ].source(),
        sourcemap:
          stats.compilation.assets[
            path.basename(entrypoint, '.cul.js') + '.js.map'
          ].source(),
      });
    });
  });
};
