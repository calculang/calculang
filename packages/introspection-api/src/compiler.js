/*
Copyright (C) 2021  DCN Consulting Ltd (incorporated in Ireland)

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License version 3,
as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

// compiler.js does a webpack pass on the given entrypoint using the webpack api (v4 atm).
// The pass is configured with loader.js defined to load calculang files (containing .cul).
// The purpose of the pass is to collect introspection information (calculang relationships) so that via graph logic
// we can determine inputs relevant to each calculang function.
// This is determined below using a graphlib Graph/s and results are stored in (also relationship inputs read from) global_state.
// webpack 'stats' object is returned/resolved below, but it doesn't actually matter: the useful output from this step is accessed by the outside world via global_state

import path from 'path';
import webpack from 'webpack';
import { createFsFromVolume, Volume } from 'memfs';

import global_state from './global_state.js';

import { Graph } from 'graphlib';
import { alg } from 'graphlib';
import dot from 'graphlib-dot';

export default (entrypoint, options = {}) => {
  var use_config = [
    {
      loader: path.resolve(__dirname, './loader.js'),
      options: { ...options, outputLocation: '' },
    },
  ];
  if (options.memo)
    // this is opt-in atm
    use_config.push({
      loader: path.resolve(__dirname, './memoloader.js'),
      options,
    });

  const compiler = webpack({
    context: __dirname,
    entry: `${path.resolve(entrypoint)}`,
    devtool: 'source-map',
    output: {
      path: path.resolve(__dirname),
      filename: 'bundle.js',
      libraryTarget: 'commonjs2', // doesn't matter
    },
    optimization: {
      minimize: false,
      //runtimeChunk: true, results in a 0.bundle.js without bootstrap, just model code bundled as jsonP
    },
    module: {
      rules: [
        {
          test: /\.cul/,
          use: use_config,
        },
        {
          test: /\.sl/, // this was an earlier development extension (seed-lang!), but keeping configured, because I still want to run my .sl files for a little while
          use: use_config,
        },
      ],
    },
  });

  compiler.outputFileSystem = createFsFromVolume(new Volume()); // will it be better to use disk directly?
  compiler.outputFileSystem.join = path.join.bind(path);

  return new Promise((resolve, reject) => {
    // note on webpack api:
    // from https://v4.webpack.js.org/api/node/:
    // "The API only supports a single concurrent compilation at a time. When using run, wait for it to finish before calling run or watch again. When using watch, call close and wait for it to finish before calling run or watch again. Concurrent compilations will corrupt the output files."
    // this is why tests are --runInBand (=> run in sequence, not in parallel)

    compiler.run((err, stats) => {
      if (err) reject(err);
      if (stats.hasErrors()) reject(stats.toJson().errors);

      debugger;

      // can I do import all logic here, and namespaces?
      // import all logic should import all things, ignoring? overwritten stuff (not explicitally exported?)

      // simple as: for all name='all' and reason='explicit import' in cul_functions
      // => loop all defns in cul_source_scope_id and create as though explicit imports. Delete 'all'
      // OK but "for all name=..." needs to be depth-first.. does it need to aff. all parents?

      // or calculate an order for all replacement in first pass... work backwards. simple implementation (no sep. tree needed)

      // 2 graphs, with edge directions reversed
      var g = new Graph({ multigraph: true });
      var g1 = new Graph({ multigraph: true });

      // create nodes: node for each cul_definition
      global_state.cul_functions.forEach((value, key) => {
        g.setNode(key, { ...value, label: `${key}` });
        g1.setNode(key, { ...value, label: `${key}` });
      });

      // set edges
      global_state.cul_links.forEach((value) => {
        if (!global_state.cul_functions.has(value.from)) return; // Array.push({}) has in cul_links as 0_undefined
        g.setEdge(value.from, value.to, {
          value,
          label:
            value.reason +
            (value.negs ? ' -' + value.negs.toString() : '') +
            (value.reason == 'input'
              ? ' +' + global_state.cul_functions.get(value.from).name
              : ''),
        });
        g1.setEdge(value.to, value.from, {
          value,
          label:
            value.reason +
            (value.negs ? ' -' + value.negs.toString() : '') +
            (value.reason == 'input'
              ? ' +' + global_state.cul_functions.get(value.from).name
              : ''),
        });
      });

      global_state.cul_functions.forEach((value, key) => {
        g.setNode(key, {
          ...value,
          //label: `${key} inputs: ${value.inputs.toString()}`,
        });
        g1.setNode(key, {
          ...value,
          //label: `${key} inputs: ${value.inputs.toString()}`,
        });
      });

      /*alg.postorder(g1, g.nodes()); // this ordering (on g1) ensures all inputs ordered before each node (all nodes included)
      alg.postorder(g1, g.nodes()).map((d) => ({
        d,
        ss: g1.successors(d).toString(),
      }));*/

      [...global_state.cul_functions.entries()]
        //.filter(([k, d]) => d.reason != 'input definition')
        .forEach(([k, d]) => {
          var a = alg;
          global_state.cul_input_map.set(
            k,
            new Set(
              alg
                .postorder(g1, k)
                .filter((d) => g.node(d)?.reason == 'input definition') // non JS has no reason e.g. reporting Table, but get excluded. Could lookup cul_functions)
                .map((d) => g.node(d).name)
            )
          );
        });

      // e.g. price_my_fave_product = () => price({product_in:'mate'}), while another call to price might indicate a dependence on product, this particular call doesn't, and *without any other calls* then price_my_fave_product should not be dependent on product
      // I call this a 'neg' in terms of the call. Logic below figures about dependence impacts via graph logic
      // cul_input_map updated directly below
      var edges_with_negs = g
        .edges()
        .map((d) => g.edge(d))
        .filter((d) => d.value.negs?.length > 0);

      edges_with_negs.forEach((edge) => {
        edge.value.negs.forEach((neg_in) => {
          var other_edges = g
            .inEdges(edge.value.to)
            .map((d) => g.edge(d))
            .filter((d) => JSON.stringify(d) != JSON.stringify(edge)); // filter out edge
          // can this forEach() be an every() with a break when a neg is needed? Also s.t. better than stringify?
          alg.preorder(g, edge.value.to).forEach((d) => {
            other_edges = g
              .inEdges(d)
              .map((d) => g.edge(d))
              .filter((d) => JSON.stringify(d) != JSON.stringify(edge))
              .filter((d) => !d.value.negs?.includes(neg_in));

            //(gg.inEdges(d).map(d => g.edge(d))

            //if (gg.inEdges(d).filter(d => g.edge(d).value.reason == 'call').map(d => g.edge(d).value.negs).reduce((a,v) => ([...a,...v]), []).includes(neg_in))
            // does another path include the neg_in?

            var result = g
              .successors(d)
              .filter(
                (d) => d != edge.value.from /* filter out the negged one */
              )
              .map((d) => global_state.cul_input_map.get(d))
              .reduce((a, v) => [...a, ...v], [])
              .includes(neg_in);

            result = other_edges
              .map((d) => global_state.cul_input_map.get(d.value.from))
              .reduce((a, v) => [...a, ...v], [])
              .includes(neg_in);

            if (!result) global_state.cul_input_map.get(d).delete(neg_in);
          });
        });
      });

      // based on cul_input_map, add inputs into cul_functions)
      global_state.cul_functions = new Map(
        [...global_state.cul_functions.entries()].map(([k, v]) => [
          k,
          {
            ...v,
            inputs: global_state.cul_input_map.has(k)
              ? [...global_state.cul_input_map.get(k)]
              : [],
          },
        ])
      );

      // update labels in the graph
      global_state.cul_functions.forEach((value, key) => {
        g.setNode(key, {
          ...value,
          label: `${key} inputs: ${value.inputs.toString()}`,
        });
        g1.setNode(key, {
          ...value,
          label: `${key} inputs: ${value.inputs.toString()}`,
        });
      });

      // put dot description of graph into global_state
      global_state.dot = dot.write(g);

      // return/resolve stats, but not used (global_state the key output)

      resolve(stats);
    });
  });
};
