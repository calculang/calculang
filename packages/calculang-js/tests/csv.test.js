/**
 * @jest-environment node
 */

import compiler from '@calculang/calculang-js';

jest.setTimeout(20000);

test.each(['manufacturing/products'])('%s', async (d) => {
  const output = await compiler(
    `./packages/calculang-testcase-models/${d}.cul.js`,
    {}
  );

  expect({
    bundle: output.bundle,
    sourcemap: output.sourcemap.replace(/\\r/g, ''), // remove carriage returns
  }).toMatchSnapshot();
});

//
