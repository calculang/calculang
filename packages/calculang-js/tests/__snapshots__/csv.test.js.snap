// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manufacturing/products 1`] = `
Object {
  "bundle": "(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define([], factory);
	else {
		var a = factory();
		for(var i in a) (typeof exports === 'object' ? exports : root)[i] = a[i];
	}
})(this, function() {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = \\"\\";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 1);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
/* harmony default export */ __webpack_exports__[\\"a\\"] = (\\"product,price\\\\n\\\\\\"solar panel\\\\\\",500\\\\n\\\\\\"heat pump\\\\\\",900\\");

/***/ }),
/* 1 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

\\"use strict\\";
__webpack_require__.r(__webpack_exports__);
/* harmony import */ var raw_loader_products_csv__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(0);
 // impt point that this doesn't contain .cul...

console.log(raw_loader_products_csv__WEBPACK_IMPORTED_MODULE_0__[/* default */ \\"a\\"]); // next -> d3-csv

/***/ })
/******/ ]);
});
//# sourceMappingURL=products.js.map",
  "sourcemap": "{\\"version\\":3,\\"sources\\":[\\"webpack:///../../../webpack/universalModuleDefinition\\",\\"webpack:///../../../webpack/bootstrap\\",\\"webpack:///../../calculang-testcase-models/manufacturing/products.csv\\",\\"webpack:///../../../products.cul.js\\"],\\"names\\":[\\"console\\",\\"log\\",\\"products_csv\\"],\\"mappings\\":\\"AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA,CAAC;AACD,O;QCVA;QACA;;QAEA;QACA;;QAEA;QACA;QACA;QACA;QACA;QACA;QACA;QACA;QACA;QACA;;QAEA;QACA;;QAEA;QACA;;QAEA;QACA;QACA;;;QAGA;QACA;;QAEA;QACA;;QAEA;QACA;QACA;QACA,0CAA0C,gCAAgC;QAC1E;QACA;;QAEA;QACA;QACA;QACA,wDAAwD,kBAAkB;QAC1E;QACA,iDAAiD,cAAc;QAC/D;;QAEA;QACA;QACA;QACA;QACA;QACA;QACA;QACA;QACA;QACA;QACA;QACA,yCAAyC,iCAAiC;QAC1E,gHAAgH,mBAAmB,EAAE;QACrI;QACA;;QAEA;QACA;QACA;QACA,2BAA2B,0BAA0B,EAAE;QACvD,iCAAiC,eAAe;QAChD;QACA;QACA;;QAEA;QACA,sDAAsD,+DAA+D;;QAErH;QACA;;;QAGA;QACA;;;;;;;;AClFe,oHAA2D,E;;;;;;;ACA1E;AAAA;CAAsD;;AAEtDA,OAAO,CAACC,GAAR,CAAYC,uEAAZ,E,CAEA,iB\\",\\"file\\":\\"products.js\\",\\"sourcesContent\\":[\\"(function webpackUniversalModuleDefinition(root, factory) {\\\\n\\\\tif(typeof exports === 'object' && typeof module === 'object')\\\\n\\\\t\\\\tmodule.exports = factory();\\\\n\\\\telse if(typeof define === 'function' && define.amd)\\\\n\\\\t\\\\tdefine([], factory);\\\\n\\\\telse {\\\\n\\\\t\\\\tvar a = factory();\\\\n\\\\t\\\\tfor(var i in a) (typeof exports === 'object' ? exports : root)[i] = a[i];\\\\n\\\\t}\\\\n})(this, function() {\\\\nreturn \\",\\" \\\\t// The module cache\\\\n \\\\tvar installedModules = {};\\\\n\\\\n \\\\t// The require function\\\\n \\\\tfunction __webpack_require__(moduleId) {\\\\n\\\\n \\\\t\\\\t// Check if module is in cache\\\\n \\\\t\\\\tif(installedModules[moduleId]) {\\\\n \\\\t\\\\t\\\\treturn installedModules[moduleId].exports;\\\\n \\\\t\\\\t}\\\\n \\\\t\\\\t// Create a new module (and put it into the cache)\\\\n \\\\t\\\\tvar module = installedModules[moduleId] = {\\\\n \\\\t\\\\t\\\\ti: moduleId,\\\\n \\\\t\\\\t\\\\tl: false,\\\\n \\\\t\\\\t\\\\texports: {}\\\\n \\\\t\\\\t};\\\\n\\\\n \\\\t\\\\t// Execute the module function\\\\n \\\\t\\\\tmodules[moduleId].call(module.exports, module, module.exports, __webpack_require__);\\\\n\\\\n \\\\t\\\\t// Flag the module as loaded\\\\n \\\\t\\\\tmodule.l = true;\\\\n\\\\n \\\\t\\\\t// Return the exports of the module\\\\n \\\\t\\\\treturn module.exports;\\\\n \\\\t}\\\\n\\\\n\\\\n \\\\t// expose the modules object (__webpack_modules__)\\\\n \\\\t__webpack_require__.m = modules;\\\\n\\\\n \\\\t// expose the module cache\\\\n \\\\t__webpack_require__.c = installedModules;\\\\n\\\\n \\\\t// define getter function for harmony exports\\\\n \\\\t__webpack_require__.d = function(exports, name, getter) {\\\\n \\\\t\\\\tif(!__webpack_require__.o(exports, name)) {\\\\n \\\\t\\\\t\\\\tObject.defineProperty(exports, name, { enumerable: true, get: getter });\\\\n \\\\t\\\\t}\\\\n \\\\t};\\\\n\\\\n \\\\t// define __esModule on exports\\\\n \\\\t__webpack_require__.r = function(exports) {\\\\n \\\\t\\\\tif(typeof Symbol !== 'undefined' && Symbol.toStringTag) {\\\\n \\\\t\\\\t\\\\tObject.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });\\\\n \\\\t\\\\t}\\\\n \\\\t\\\\tObject.defineProperty(exports, '__esModule', { value: true });\\\\n \\\\t};\\\\n\\\\n \\\\t// create a fake namespace object\\\\n \\\\t// mode & 1: value is a module id, require it\\\\n \\\\t// mode & 2: merge all properties of value into the ns\\\\n \\\\t// mode & 4: return value when already ns object\\\\n \\\\t// mode & 8|1: behave like require\\\\n \\\\t__webpack_require__.t = function(value, mode) {\\\\n \\\\t\\\\tif(mode & 1) value = __webpack_require__(value);\\\\n \\\\t\\\\tif(mode & 8) return value;\\\\n \\\\t\\\\tif((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;\\\\n \\\\t\\\\tvar ns = Object.create(null);\\\\n \\\\t\\\\t__webpack_require__.r(ns);\\\\n \\\\t\\\\tObject.defineProperty(ns, 'default', { enumerable: true, value: value });\\\\n \\\\t\\\\tif(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));\\\\n \\\\t\\\\treturn ns;\\\\n \\\\t};\\\\n\\\\n \\\\t// getDefaultExport function for compatibility with non-harmony modules\\\\n \\\\t__webpack_require__.n = function(module) {\\\\n \\\\t\\\\tvar getter = module && module.__esModule ?\\\\n \\\\t\\\\t\\\\tfunction getDefault() { return module['default']; } :\\\\n \\\\t\\\\t\\\\tfunction getModuleExports() { return module; };\\\\n \\\\t\\\\t__webpack_require__.d(getter, 'a', getter);\\\\n \\\\t\\\\treturn getter;\\\\n \\\\t};\\\\n\\\\n \\\\t// Object.prototype.hasOwnProperty.call\\\\n \\\\t__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };\\\\n\\\\n \\\\t// __webpack_public_path__\\\\n \\\\t__webpack_require__.p = \\\\\\"\\\\\\";\\\\n\\\\n\\\\n \\\\t// Load entry module and return exports\\\\n \\\\treturn __webpack_require__(__webpack_require__.s = 1);\\\\n\\",\\"export default \\\\\\"product,price\\\\\\\\\\\\n\\\\\\\\\\\\\\"solar panel\\\\\\\\\\\\\\",500\\\\\\\\\\\\n\\\\\\\\\\\\\\"heat pump\\\\\\\\\\\\\\",900\\\\\\";\\",\\"import products_csv from 'raw-loader!./products.csv'; // impt point that this doesn't contain .cul...\\\\n\\\\nconsole.log(products_csv);\\\\n\\\\n// next -> d3-csv\\\\n\\"],\\"sourceRoot\\":\\"\\"}",
}
`;
